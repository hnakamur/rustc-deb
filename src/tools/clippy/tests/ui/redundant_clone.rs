#![warn(clippy::redundant_clone)]

use std::ffi::OsString;
use std::path::Path;

fn main() {
    let _ = ["lorem", "ipsum"].join(" ").to_string();

    let s = String::from("foo");
    let _ = s.clone();

    let s = String::from("foo");
    let _ = s.to_string();

    let s = String::from("foo");
    let _ = s.to_owned();

    let _ = Path::new("/a/b/").join("c").to_owned();

    let _ = Path::new("/a/b/").join("c").to_path_buf();

    let _ = OsString::new().to_owned();

    let _ = OsString::new().to_os_string();

    // Check that lint level works
    #[allow(clippy::redundant_clone)]
    let _ = String::new().to_string();

    let tup = (String::from("foo"),);
    let _ = tup.0.clone();

    let tup_ref = &(String::from("foo"),);
    let _s = tup_ref.0.clone(); // this `.clone()` cannot be removed
}

#[derive(Clone)]
struct Alpha;
fn with_branch(a: Alpha, b: bool) -> (Alpha, Alpha) {
    if b {
        (a.clone(), a.clone())
    } else {
        (Alpha, a)
    }
}

struct TypeWithDrop {
    x: String,
}

impl Drop for TypeWithDrop {
    fn drop(&mut self) {}
}

fn cannot_move_from_type_with_drop() -> String {
    let s = TypeWithDrop { x: String::new() };
    s.x.clone() // removing this `clone()` summons E0509
}
