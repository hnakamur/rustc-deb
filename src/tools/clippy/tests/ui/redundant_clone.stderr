error: redundant clone
  --> $DIR/redundant_clone.rs:7:41
   |
LL |     let _ = ["lorem", "ipsum"].join(" ").to_string();
   |                                         ^^^^^^^^^^^^ help: remove this
   |
   = note: `-D clippy::redundant-clone` implied by `-D warnings`
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:7:13
   |
LL |     let _ = ["lorem", "ipsum"].join(" ").to_string();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: redundant clone
  --> $DIR/redundant_clone.rs:10:14
   |
LL |     let _ = s.clone();
   |              ^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:10:13
   |
LL |     let _ = s.clone();
   |             ^

error: redundant clone
  --> $DIR/redundant_clone.rs:13:14
   |
LL |     let _ = s.to_string();
   |              ^^^^^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:13:13
   |
LL |     let _ = s.to_string();
   |             ^

error: redundant clone
  --> $DIR/redundant_clone.rs:16:14
   |
LL |     let _ = s.to_owned();
   |              ^^^^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:16:13
   |
LL |     let _ = s.to_owned();
   |             ^

error: redundant clone
  --> $DIR/redundant_clone.rs:18:41
   |
LL |     let _ = Path::new("/a/b/").join("c").to_owned();
   |                                         ^^^^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:18:13
   |
LL |     let _ = Path::new("/a/b/").join("c").to_owned();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: redundant clone
  --> $DIR/redundant_clone.rs:20:41
   |
LL |     let _ = Path::new("/a/b/").join("c").to_path_buf();
   |                                         ^^^^^^^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:20:13
   |
LL |     let _ = Path::new("/a/b/").join("c").to_path_buf();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: redundant clone
  --> $DIR/redundant_clone.rs:22:28
   |
LL |     let _ = OsString::new().to_owned();
   |                            ^^^^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:22:13
   |
LL |     let _ = OsString::new().to_owned();
   |             ^^^^^^^^^^^^^^^

error: redundant clone
  --> $DIR/redundant_clone.rs:24:28
   |
LL |     let _ = OsString::new().to_os_string();
   |                            ^^^^^^^^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:24:13
   |
LL |     let _ = OsString::new().to_os_string();
   |             ^^^^^^^^^^^^^^^

error: redundant clone
  --> $DIR/redundant_clone.rs:31:18
   |
LL |     let _ = tup.0.clone();
   |                  ^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:31:13
   |
LL |     let _ = tup.0.clone();
   |             ^^^^^

error: redundant clone
  --> $DIR/redundant_clone.rs:41:22
   |
LL |         (a.clone(), a.clone())
   |                      ^^^^^^^^ help: remove this
   |
note: this value is dropped without further use
  --> $DIR/redundant_clone.rs:41:21
   |
LL |         (a.clone(), a.clone())
   |                     ^

error: aborting due to 10 previous errors

